audio-worklet/design-pattern/lib/wa2.js

```diff
@@ -0,0 +1,289 @@
+/**
+ * @fileOverview WA2 library. A set of utilities to support WASM operations
+ * for the WebAudio.
+ */
+
+
+/**
+ * @namespace WA2
+ */
+var WA2 = {
+  // Basic byte unit of WASM heap. (16 bit = 2 bytes)
+  BYTES_PER_UNIT: Uint16Array.BYTES_PER_ELEMENT,
+
+  // Byte per audio sample. (32 bit float)
+  BYTES_PER_SAMPLE: Float32Array.BYTES_PER_ELEMENT,
+
+  // The max audio channel on Chrome is 32.
+  MAX_CHANNEL_COUNT: 32,
+
+  // WebAudio's render quantum size.
+  RENDER_QUANTUM_FRAMES: 128,
+};
+
+
+/**
+ * Copy AudioBuffer by a specified channel count.
+ * @param  {AudioBuffer} sourceBuffer
+ * @param  {AudioBuffer} destinationBuffer
+ * @param  {number} channelCount A number of channels to be copied.
+ */
+WA2.copyAudioBuffer = function(sourceBuffer, destinationBuffer, channelCount) {
+  console.assert(channelCount <= sourceBuffer.numberOfChannels);
+  console.assert(channelCount <= destinationBuffer.numberOfChannels);
+
+  for (let i = 0; i < channelCount; ++i) {
+    destinationBuffer.getChannelData(i).set(sourceBuffer.getChannelData(i));
+  }
+};
+
+
+/**
+ * A WASM HEAP wrapper for AudioBuffer class. This breaks down the AudioBuffer
+ * into an Array of Float32Array for the convinient WASM opearion.
+ *
+ * @class
+ * @dependency Module A WASM module generated by the emscripten glue code.
+ */
+class HeapAudioBuffer {
+
+  /**
+   * @constructor
+   * @param  {number} length Buffer frame length.
+   * @param  {number} channelCount Number of channels.
+   * @param  {number=} maxChannelCount Maximum number of channels.
+   * @return {object} HeapAudioBuffer object.
+   */
+  constructor(length, channelCount, maxChannelCount) {
+    console.assert(Module);
+    console.assert(0 < channelCount && channelCount <= WA2.MAX_CHANNEL_COUNT);
```

Comment:
Need to check length and maxChannelCount too?

Reply:
Well, it's just that you have several asserts, but don't verify that length and maxChannelCount make sense.  So unless this is guaranteed in some other way, it seems natural for asserts for them as well.